### ISSUE TITLE
Github Username Change should not prevent user from Logging in to DEV

### ISSUE THREAD

**Issue by mstruve** [forem/forem#5268]
**Describe the bug**
When a user changes their name on Github the UID that we get from github for authentication seems to change as well. This means that if a user signs up for DEV, changes their github username, and then tries to sign back into DEV they will get a "User has already been taken error"

The reason for this is because we try to [look up their identity using that UID](https://github.com/thepracticaldev/dev.to/blob/master/app/models/identity.rb#L11):
```ruby
def self.find_for_oauth(auth)
  find_or_create_by(uid: auth.uid, provider: auth.provider)
end
``` 
and when we don't find it we assume that the user is new and we try to create a new one. We try to prevent this scenario [HERE](https://github.com/thepracticaldev/dev.to/blob/master/app/services/authorization_service.rb#L49):
```ruby
user = User.where("#{identity.provider}_username" => auth.info.nickname).first
```
but that doesnt work and will fail to find the user bc the username in our system is the old username. 

We need to find a way to recognize and connect a Github user with it's corresponding DEV user after a user has changed their Github username.

**Comment by michael-tharrington** [comment]
Thanks for writing this up!

Folks write in about this error message pretty often and while we do have a manual way for admins to fix this, it'd be great for the system to handle it.

To elaborate, users are hit with this error message:

**Something went wrong: Log in error: Validation failed: User has already been taken
Please try again below, or contact yo@dev.to if this persists.**

When they contact us, we have a tool in `internal/users` that allows us to check if a user's GitHub or Twitter UID's don't match their identity's UID. Assuming they don't match, we use the tool to delete their ID. This allows them to log into their account. We then give them instruction to go to their settings - https://dev.to/settings/account and reconnect the removed Twitter or GitHub account.

This tool is also useful if a user deletes their GitHub or Twitter account — in other words, it's not just for solving cases of renamed accounts.

**Comment by citizen428** [comment]
tl;dr: While we store the unchangeable numeric id, we use the changeable username to determine if a user is new.

Longer version:

From a quick look at my own account in the dev environment, the stored `Identity` makes sense, since it uses the `id` attribute of the `Omniauth::Github` auth hash to populate the `uid` column:

```
"provider" => "github",
 "uid" => "<numeric id as string>"
```

The problem is that the `where` clause trying to determine if the user is new relies on the name instead of the id, which has changed, so the user is considered new. Trying to create the new user then triggers the uniqueness validations in the `Identity` class.

This *should* fix the problem from some quick testing in the Rails console:

```ruby
# before
user = User.where("#{identity.provider}_username" => auth.info.nickname).first

# after
provider_uid = auth.extra.raw_info.id.to_s
user = User.includes(:identities).where(identities: { uid: provider_uid}).first
```

@mstruve @benhalpern If you want me to look at this in more detail, feel free to assign me this issue and ping me on Slack, I'll pick it up tomorrow.

**Comment by mstruve** [comment]
@citizen428 The first thing we do before trying the username is [do the lookup using the provider uid](https://github.com/thepracticaldev/dev.to/blob/master/app/models/identity.rb#L11) however, the uid changes which is why that initial lookup fails and we fallback on trying to use the username.

**Comment by citizen428** [comment]
@mstruve The numerical user id changes when you change your GH username? That seems a bit odd.

**Comment by citizen428** [comment]
It seems somewhat unlikely that the numeric ID would change when the username changes. 

I just ran through the following:

1. Create a new dummy Github account for testing
2. Sign up through Github in my dev environment. All data looks correct:
    ```ruby
    u.username
    #=> "dsfsfad"
   i = u.identities.first
   i.uid
    #=> "59xxxxxx"
   i.auth_data_dump.dig("extra", "raw_info", "id")
   #=> 59xxxxxx
   i.auth_data_dump.dig("info", "nickname")
   "dsfsfad"
   ```
3. Sign out from DEV
4. Change username on Github to "dsfsfad1".
5. Sign back into DEV. All successful. The user's data now looks like this:
    ```ruby
    u.identities.count
    #=> 1
    u.username
    #=> "dsfsfad"
    i = u.identities.first
    i.uid
    #=> "59xxxxxx"
    # unchanged!
   i.auth_data_dump.dig("extra", "raw_info", "id") 
   #=> 59xxxxxx
   # unchanged!
   i.auth_data_dump.dig("info", "nickname")
   "dsfsfad1"
   # changed!
   ```
6. Change the username again on GH again, this time without logging out first.
7. Log out and back in again. Same as before, the UID stayed the same, the nickname stored in `info` got updated to "dsfsfad2".

This seems to corroborate the idea that the UID stays stable across account renames. Putting a `binding.pry` inside `Identity.find_for_oauth` also seems to confirm this, despite repeated name changes it returns the correct identity.

Of course this doesn't get closer to the root cause yet, but it seems like this only happens under some circumstances and not others.

**Comment by mstruve** [comment]
Thanks for running that test @citizen428! I guess the Github username change might be a red herring then and the UID changes for some other reason. I ran into this with @avery-d and I can confirm her UID had changed from when she signed up in Sept. I noticed her Github username had also changed and assumed that might have been why. It was likely due to some other setting however.  

@avery-d did you make any other Github changes to your account after you used it to sign in to DEV?

**Comment by citizen428** [comment]
> When they contact us, we have a tool in internal/users that allows us to check if a user's GitHub or Twitter UID's don't match their identity's UID. Assuming they don't match, we use the tool to delete their ID. 

@michael-tharrington Just to make sure we're all talking about the same thing: When you say UID here, do you mean numeric IDs or usernames?

**Comment by mstruve** [comment]
Given this has not occurred recently I am going to close.
